File: python/Q125_valid_palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#


# @lc code=start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        formated: list[str] = [c for c in s.lower() if c.isalnum()]

        left: int = 0
        right: int = len(formated) - 1

        while left < right:
            if formated[left] != formated[right]:
                return False

            left += 1
            right -= 1

        return True


# @lc code=end

File: rust/src/leetcode/Q125_valid_palindrome.rs
/*
 * @lc app=leetcode id=125 lang=rust
 *
 * [125] Valid Palindrome
 */

// @lc code=start
impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        let chars: Vec<char> = s
            .to_lowercase()
            .chars()
            .filter(|c| c.is_alphanumeric())
            .collect();

        if chars.is_empty() {
            return true;
        }

        let mut left: usize = 0;
        let mut right: usize = chars.len() - 1;

        while left < right {
            if chars[left] != chars[right] {
                return false;
            }

            left += 1;
            right -= 1;
        }

        true
    }
}
// @lc code=end
